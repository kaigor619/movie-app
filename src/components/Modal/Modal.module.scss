.modalWrap {
  display: flex;
  align-items: center;
  justify-content: center;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100vh;
}

.modalContent {
  position: relative;
  z-index: 1050;
  max-height: 100%;
  width: 100%;
  max-width: 28rem;
}

.background {
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  z-index: 1000;
  display: block;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  outline: 0;
}

.enter {
  .modalContent {
    opacity: 0;
    transform: scale(0.9);
  }

  .background {
    opacity: 0;
  }
}

.enterDone {
  .modalContent {
    opacity: 1;
    transform: scale(1);
    transition: all 0.15s;
  }

  .background {
    opacity: 1;
    transition: all 0.15s;
  }
}

.exit {
  .modalContent {
    opacity: 0;
    transform: scale(0.9);
    transition: all 0.2s;
  }

  .background {
    opacity: 0;
    transition: opacity 0.2s;
  }
}
