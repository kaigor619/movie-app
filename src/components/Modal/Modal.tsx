import React, { useRef } from "react";
import cls from "clsx";
import ReactDOM from "react-dom";
import { CSSTransition } from "react-transition-group";
import styles from "./Modal.module.scss";

interface Props {
  open: boolean;
  onClose: () => void;
  children: React.ReactNode | React.ReactNode[];
  maxWidth: number;
}

export const Modal: React.FC<Props> = ({
  open,
  onClose,
  children,
  maxWidth,
}) => {
  const modalRootRef = useRef(document.getElementById("modal-root"));

  if (!modalRootRef.current) return null;

  return ReactDOM.createPortal(
    <CSSTransition
      in={open}
      timeout={{
        exit: 200,
      }}
      classNames={{
        enter: styles.enter,
        enterDone: styles.enterDone,
        exit: styles.exit,
      }}
      unmountOnExit
    >
      <div className={styles.modalWrap}>
        <div
          data-testid="overlay"
          className={styles.background}
          onMouseDown={onClose}
        />
        <div
          data-testid="dialog"
          className={cls(styles.modalContent)}
          style={{ maxWidth: `${maxWidth}rem` }}
        >
          {children}
        </div>
      </div>
    </CSSTransition>,
    modalRootRef.current
  );
};
