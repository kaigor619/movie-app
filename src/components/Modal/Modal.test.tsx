import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Modal } from "./Modal";

describe("Modal component", () => {
  const createModalDirectory = () => {
    if (!document.getElementById("modal-root")) {
      const modalRoot = document.createElement("div");

      modalRoot.id = "modal-root";

      document.body.appendChild(modalRoot);
    }
  };

  const onClose = jest.fn();

  afterEach(() => {
    onClose.mockClear();
  });

  test("shows null if modal-root not found", async () => {
    render(
      <Modal open maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    expect(screen.queryByText(/Hello world!/i)).not.toBeInTheDocument();
  });

  it("shows Modal content", async () => {
    createModalDirectory();

    render(
      <Modal open maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    await waitFor(() =>
      expect(screen.getByText(/Hello world!/i)).toBeInTheDocument()
    );
  });
  it("checks modal width", async () => {
    createModalDirectory();

    render(
      <Modal open maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    await waitFor(() =>
      expect(screen.getByTestId("dialog")).toHaveStyle({ maxWidth: "30rem" })
    );
  });

  it("checks snapshots", async () => {
    createModalDirectory();
    const { baseElement } = render(
      <Modal open maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    await waitFor(() => expect(screen.getByText(/Hello/i)).toBeInTheDocument());

    expect(baseElement).toMatchSnapshot();
  });

  it("fires onClose when click on overlay", async () => {
    createModalDirectory();

    render(
      <Modal open maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    await waitFor(() => expect(screen.getByText(/Hello/i)).toBeInTheDocument());

    await userEvent.click(screen.getByTestId("overlay"));

    expect(onClose).toBeCalledTimes(1);
  });

  test("closes Modal via updating open prop", async () => {
    createModalDirectory();

    const { rerender, unmount } = render(
      <Modal open maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    await waitFor(() =>
      expect(screen.getByText(/Hello world!/i)).toBeInTheDocument()
    );

    rerender(
      <Modal open={false} maxWidth={30} onClose={onClose}>
        <h1>Hello world!</h1>
      </Modal>
    );

    await waitFor(() =>
      expect(screen.queryByTestId("dialog")).not.toBeInTheDocument()
    );
  });
});
