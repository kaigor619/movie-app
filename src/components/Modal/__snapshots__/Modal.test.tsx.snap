// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component checks snapshots 1`] = `
<body>
  <div
    id="modal-root"
  >
    <div
      class="modalWrap"
    >
      <div
        class="background"
        data-testid="overlay"
      />
      <div
        class="modalContent"
        data-testid="dialog"
        style="max-width: 30rem;"
      >
        <h1>
          Hello world!
        </h1>
      </div>
    </div>
  </div>
  <div />
</body>
`;
